custom/mgh_epic/sync_epic_appointments.py

```diff
@@ -0,0 +1,299 @@
+import uuid
+
+from datetime import datetime
+from django.conf import settings
+
+from jsonpath_ng import parse as parse_jsonpath
+from urllib.parse import urlencode
+
+from corehq.apps.hqcase.case_helper import CaseHelper
+from corehq.form_processor.models import CommCareCase, CommCareCaseIndex
+from corehq.util.timezones.conversions import ServerTime
+
+
+OAUTH_URL = "https://ws-interconnect-fhir.partners.org/Interconnect-FHIR-MU-PRD/oauth2/token"
+PATIENT_URL = "https://ws-interconnect-fhir.partners.org/Interconnect-FHIR-MU-PRD/api/FHIR/R4/Patient"
+APPOINTMENT_URL = "https://ws-interconnect-fhir.partners.org/Interconnect-FHIR-MU-PRD/api/FHIR/R4/Appointment"
+
+
+def handle_response(response):
+    if 200 <= response.status_code < 300:
+        return response.json()
+    response.raise_for_status()
+
+
+def generate_epic_jwt():
+    key = settings.EPIC_PRIVATE_KEY
+    # token will expire in 4 mins
+    exp = int(time.time()) + 240
+    jti = str(uuid.uuid4())
+    header = {
+        "alg": "RS256",
+        "typ": "JWT",
+    }
+    payload = {
+        "iss": settings.EPIC_CLIENT_ID,
+        "sub": settings.EPIC_CLIENT_ID,
+        "aud": OAUTH_URL,
+        "jti": jti,
+        "exp": exp
+    }
+    token = jwt.encode(payload, key, algorithm="RS256", headers=header)
+    return token
+
+
+def request_epic_access_token():
+    headers = {
+        "Content_Type": "application/x-www-form-urlencoded",
+    }
+    data = {
+        "grant_type": "client_credentials",
+        "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+        "client_assertion": generate_epic_jwt()
+    }
+    url = OAUTH_URL
+    response = requests.post(url, data=data, headers=headers)
+    response_json = handle_response(response)
+
+    return response_json.get('access_token')
+
+
+def get_patient_fhir_id(given_name, family_name, birthdate, access_token):
+    base_url = PATIENT_URL
+    params = {
+        'birthdate': birthdate,
+        'family': family_name,
+        'given': given_name,
+        '_format': 'json'
+    }
+    url = f'{base_url}?{urlencode(params)}'
+    headers = {
+        'authorization': f'Bearer {access_token}',
+    }
+    response = requests.get(url, headers=headers)
+    response_json = handle_response(response)
+    fhir_id = None
+
+    path = parse_jsonpath('entry[0].resource.id')
+    matches = path.find(response_json)
+    if matches:
+        fhir_id = matches[0].value
+    return fhir_id
+
+
+def filter_appointments_by_date(appointments_entries_json, study_start_date, weeks_from_start_date):
+    """
+    Returns a list of appointment entries
+    :param appointments_entries_json: Json response data for Epic appointment search
+    :param study_start_date: A date case property formatted YYYY-MM-DD
+    :param weeks_from_start_date: Number of weeks after study start to filter by
+    :return: A list of json entries
+    """
+    appointments = []
+    path = parse_jsonpath('$.resource.start')
```

Comment:
jsonpath feels a bit heavy for something simple like this: `entry.get("resource", {}).get("start")`

Reply:
also part of https://github.com/dimagi/commcare-hq/commit/6290acdb30a5f2329650ceb5ec255eb1195d73f0